### PR TITLE
feat: natspec, fixes and tests

### PR DESCRIPTION
**Pull request by md0x**
Signed-off-by: Pablo Maldonado <pablo@umaproject.org>

Changes proposed in this price request:
- Simplify final fee update logic in Internal Optimistic Oracle
- Internal Optimistic Oracle nat spec
- Internal Optimistic Oracle edge test cases
- Small fixes in tests

### CODE REVIEW

**Review by chrismaree — APPROVED** [review]

File: contracts/InternalOptimisticOracle.sol [hunk 1]
```diff
@@ -47,8 +63,15 @@ contract InternalOptimisticOracle is Testable {
...
 
-    mapping(bytes32 => Request) public requests;
-
+    /**
+     * @notice Requests a price from the Internal Optimistic Oracle.
+     * @param timestamp timestamp of the price being requested.
+     * @param ancillaryData ancillary data of the price being requested.
+     * @param reward reward offered to a successful proposer. Will be pulled from the caller. Note: this can be 0, which could
+     * make sense if the contract requests and proposes the value in the same call or provides its own reward system.
+     * @param bond custom proposal bond to set for request. If set to 0, defaults to the final fee.
```
Comment (Reinis-FRP) [inline contracts/InternalOptimisticOracle.sol:72]: Does it really set bond to final fee if 0? `oo.setBond` in `disputePrice` would explicitly set it to 0

File: contracts/InternalOptimisticOracle.sol [hunk 2]
```diff
@@ -103,32 +140,37 @@ contract InternalOptimisticOracle is Testable {
 
         request.disputer = msg.sender;
 
-        // If the final fee gets updated between the time the request is made and the time the dispute is made, the
+        // If the final fee gets increased between the time the request is made and the time the dispute is made, the
```
Comment (Reinis-FRP) [inline contracts/InternalOptimisticOracle.sol:143]: I might have been wrong in my previous comments on this edge case as rereading OO I see that `finalFee` gets stored on price request.

File: contracts/InternalOptimisticOracle.sol (hunk 1 above)
Comment (md0x) [reply contracts/InternalOptimisticOracle.sol:72]: We always call [oo.setBond](https://github.com/UMAprotocol/dev-quickstart/blob/c5d8a9148546e641d289cafd06e80bd8f434c35e/contracts/InternalOptimisticOracle.sol#L161) in the dispute, so if we pass bond=0, both `oo.proposePriceFo` and `oo.disputePriceFor` will pull a `finalFee` amount of currency each, `2*finalFee` in total

How would you phrase it?

File: contracts/InternalOptimisticOracle.sol (hunk 2 above)
Comment (md0x) [reply contracts/InternalOptimisticOracle.sol:143]: I think you were right last time when you thought about this:
In  this contract we have a local [RequestPrice](https://github.com/UMAprotocol/dev-quickstart/blob/c5d8a9148546e641d289cafd06e80bd8f434c35e/contracts/InternalOptimisticOracle.sol#L27) object instantiated when we do the local `requestPrice`. During this call we store the final fee.

If the `Store.finalFee` is updated before anyone calls `OA.proposePrice` or `OA.disputePrice` the local `OA.request.finalFee` would be outdated and then won't work when doing the dispute. Does it make sense?

**Pull request merged**